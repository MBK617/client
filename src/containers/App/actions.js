export const setToken = (token) => ({ type: 'SET_TOKEN', token });
export const setTokenSuccess = (token) => ({ type: 'SET_TOKEN_SUCCESS', token });
export const getToken = () => ({ type: 'GET_TOKEN' });
export const getTokenSuccess = (token) => ({ type: 'GET_TOKEN_SUCCESS', token });

export const pushHistory = (path) => ({ type: 'PUSH_HISTORY', path });
export const popHistory = () => ({ type: 'POP_HISTORY' });