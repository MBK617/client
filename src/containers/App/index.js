import React from 'react';
import { View } from 'react-native';

import Counter from 'containers/Counter';

import styles from './styles';

function App() {
  return (
    <View style={styles.container}>
      <Counter/>
    </View>
  );
}

export default App;