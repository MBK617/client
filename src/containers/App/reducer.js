/*
 * AppReducer
 */

import { fromJS } from 'immutable';

export const initialState = fromJS({
  token: null,
}).set('history', ['home']);

function appReducer(state = initialState, action) {
  switch (action.type) {
    case 'GET_TOKEN_SUCCESS':
      return state.set('token', action.token);
    case 'SET_TOKEN_SUCCESS':
      return state.set('token', action.token);
    case 'PUSH_HISTORY':
      return state.set('history', [action.path, ...state.get('history')]);
    case 'POP_HISTORY':
      return state.set('history', state.get('history').slice(1));
    default:
      return state;
  }
}

export default appReducer;