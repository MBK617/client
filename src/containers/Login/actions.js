export const logIn = (email, password) => ({ type: 'LOGIN', email, password })
export const logInSuccess = (token) => ({ type: 'LOGIN_SUCCESS', token })
export const logInFailure = (error) => ({ type: 'LOGIN_FAILURE', error })