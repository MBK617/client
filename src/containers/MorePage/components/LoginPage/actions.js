export const logIn = (email, password) => ({ type: 'LOGIN', email, password })
export const logInSuccess = () => ({ type: 'LOGIN_SUCCESS' })
export const logInFailure = (error) => ({ type: 'LOGIN_FAILURE', error })