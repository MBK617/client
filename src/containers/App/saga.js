import { all, takeLatest, put } from "redux-saga/effects";
import * as storage from 'utils/storage';
import { getTokenSuccess, setTokenSuccess } from "./actions";

export function* setTokenInStorage(action) {
  try {
    if(action.token) {
      yield storage.setItem('token', action.token);
    } else {
      yield storage.deleteItem('token');
    }
    yield put(setTokenSuccess(action.token));
  } catch (err) {
    console.error(err);
  }
}

export function* getTokenFromStorage() {
  try {
    const value = yield storage.getItem('token');
    if(value !== null) {
      yield put(getTokenSuccess(value));
    }
  } catch(err) {
    console.error(err);
  }
}

export default function* rootSaga() {
  yield all([
    takeLatest('SET_TOKEN', setTokenInStorage),
    takeLatest('GET_TOKEN', getTokenFromStorage),
  ]);
}