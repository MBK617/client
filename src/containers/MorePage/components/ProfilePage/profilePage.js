import React from 'react';
import { View, Text, Image, Button } from "react-native";
import { useDispatch } from 'react-redux';

import { setToken } from 'containers/App/actions';

import styles from './styles';

const ProfilePage = () => {
  const dispatch = useDispatch();
  return (
    <View style={styles.root}>
      <Text>You are logged in!</Text>
      <Image style={styles.image} source = {{ uri: 'https://i.redd.it/tonbtar3mb741.jpg' }}/>
      <Button title='Log Out' onPress={()=>dispatch(setToken(null))}/>
    </View>
  );
}

export default ProfilePage;