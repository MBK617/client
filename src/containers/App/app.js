import React, { useEffect } from 'react';
import { View, Image } from 'react-native';
import { compose } from 'redux';

import { 
  Appbar, 
  BottomNavigation 
} from 'react-native-paper';

import injectSaga from 'utils/redux/injectSaga';
import injectReducer from 'utils/redux/injectReducer';

import useHistory from 'hooks/useHistory';
import useLoginStatus from 'hooks/useLoginStatus';

import styles from './styles';
import reducer from './reducer';
import saga from './saga';

import HomePage from '../HomePage';
import EventsPage from '../EventsPage';
import InboxPage from '../InboxPage';
import MorePage from '../MorePage';

const pages = [
  { key: 'home', icon: 'home', authenticated: false },
  { key: 'events', icon: 'calendar', authenticated: true },
  { key: 'inbox', icon: 'forum', authenticated: true },
  { key: 'more', icon: 'dots-horizontal', authenticated: false },
]

const App = () => {
  const isLoggedIn = useLoginStatus();
  const { history, path, goTo, goBack } = useHistory();

  const [index, setIndex] = React.useState(0);
  const [routes, setRoutes] = React.useState(pages.filter(({authenticated}) => isLoggedIn || !authenticated).map(({ authenticated, ...page }) => page));

  useEffect(() => {
    const newRoutes = pages.filter(({authenticated}) => isLoggedIn || !authenticated).map(({ authenticated, ...page }) => page);
    setIndex(newRoutes.findIndex(route => route.key === routes[index].key))
    setRoutes(newRoutes);
  }, [isLoggedIn])

  useEffect(() => {
    const newPath = routes[index].key;
    goTo(newPath);
  }, [index])

  useEffect(() => {
    const newIndex = routes.findIndex(page => page.key === path?.split('/')[0]);
    setIndex(newIndex);
  }, [path])

  const renderScene =  () => {
    switch (path?.split('/')[0]) {
      case 'home':
        return <HomePage />;
      case 'events':
        return <EventsPage />;
      case 'inbox':
        return <InboxPage />;
      case 'more':
        return <MorePage />;
    }
  }
  
  return (
    <View style={styles.root}>
      <Appbar.Header 
        statusBarHeight={0}
        style={styles.header}
      > 
        <Appbar.BackAction onPress={goBack} style={{opacity: history.length > 1 ? 1 : 0, transition: '.2s'}}/>
        <Image source={require('assets/mbklogo.jpg')} style={styles.logo}/>
        <Appbar.Action icon="bell" />
      </Appbar.Header>
      <BottomNavigation
        barStyle={{ backgroundColor: '#222' }}
        navigationState={{ index, routes }}
        onIndexChange={setIndex}
        renderScene={renderScene}
        shifting={false} 
        labeled={false} 
      />
    </View>
  );
}

const withReducer = injectReducer({ key: 'app', reducer });
const withSaga = injectSaga({ key: 'app', saga });

export default compose(withReducer, withSaga)(App);