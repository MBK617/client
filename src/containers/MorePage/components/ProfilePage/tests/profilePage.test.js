import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';
import { fromJS } from 'immutable';

import { setToken } from 'containers/App/actions';

import ProfilePage from '..';

const initialState = {
  app: fromJS({
    token: '',
  })
};

describe('<ProfilePage />', () => {
  const sagaMiddleware = createSagaMiddleware();
  let store, rendered;
  
  const mockStore = (initialState) => ({ 
    ...configureStore([sagaMiddleware])(initialState), 
    injectedReducers: {}, 
    injectedSagas: {}, 
    runSaga: sagaMiddleware.run 
  });
  
  beforeEach(() => {
    store = mockStore(initialState);
    rendered = render(<Provider store={store}><ProfilePage /></Provider>);
  })

  afterEach(cleanup);

  it('should render correctly', () => {
    const tree = rendered.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('log out button should dispatch log out action', () => {
    fireEvent.press(rendered.getByText(/Log Out/i));
    const actions = store.getActions();
    expect(actions[0]).toEqual(setToken(null));
  });
});