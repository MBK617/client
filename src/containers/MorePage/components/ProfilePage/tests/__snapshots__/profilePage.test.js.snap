// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfilePage /> should render correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexShrink": 0,
      "justifyContent": "center",
      "minWidth": 250,
      "padding": 10,
      "width": "100%",
    }
  }
>
  <Text>
    You are logged in!
  </Text>
  <Image
    source={
      Object {
        "uri": "https://i.redd.it/tonbtar3mb741.jpg",
      }
    }
    style={
      Object {
        "height": 200,
        "margin": 10,
      }
    }
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Log Out
      </Text>
    </View>
  </View>
</View>
`;
